write a test for projects -> init

This one should be easier than others to start, which I am basing off of the notion of 'total HDK function calls'

First, add `pub` to it so you can reference it outside the crate, in the tests folder. 
Create a new test file that corresponds to this `lib.rs` file, over in `tests` crate folder.
Writing the test will require drilling into the two function calls, then looking for calls to the HDK that will need to be mocked


https://github.com/h-be/acorn/blob/7fab3552fad60036eca4dd8f8d5b2aebbdd45139/dna/zomes/projects/src/lib.rs#L17-L27

@Connoropolous, I've hit a bit of a roadblock, whenever I try to run a test (even if it's one you've already written), I get the following error:
```
error[E0658]: use of unstable library feature 'duration_constants'
   --> /home/wesley/.cargo/registry/src/github.com-1ecc6299db9ec823/kitsune_p2p-0.0.3/src/gossip/sharded_gossip.rs:377:41
    |
377 |         const START_OF_TIME: Duration = Duration::MAX;
    |                                         ^^^^^^^^^^^^^
    |
    = note: see issue #57391 <https://github.com/rust-lang/rust/issues/57391> for more information

error: aborting due to previous error

For more information about this error, try `rustc --explain E0658`.
error: could not compile `kitsune_p2p`
```

Am I working with the wrong version of the code?

--- a/dna/tests/Cargo.toml
+++ b/dna/tests/Cargo.toml
@@ -14,6 +14,7 @@ paste = "*"
 serde = "*"
 thiserror = "1"
 mockall = "0.9"
+assert_matches = "1.5.0"
 
 [dev-dependencies]
 fixt = "0.0.5"

--- a/dna/tests/src/project/mod.rs
+++ b/dna/tests/src/project/mod.rs
@@ -7,3 +7,4 @@ pub mod goal_vote;
 pub mod member;
 pub mod project_meta;
 pub mod fixtures;
+pub mod project_init;

--- a/dna/tests/src/project/project_init.rs
+++ b/dna/tests/src/project/project_init.rs
@@ -1,0 +1,84 @@
+#[cfg(test)]
+pub mod tests {
+    use crate::project::fixtures::fixtures::ProjectMetaFixturator;
+    use ::fixt::prelude::*;
+    use hdk::prelude::*;
+    use hdk_crud::WrappedAgentPubKey;
+    use holochain_types::prelude::option_entry_hashed;
+    use holochain_types::prelude::ElementFixturator;
+    use holochain_types::prelude::ValidateDataFixturator;
+    use projects::project::error::Error;
+    use projects::project::project_meta::crud::ProjectMeta;
+    use projects::project::project_meta::validate::*;
+    use projects::init;
+
+    #[test]
+    fn test_init() {
+               
+        let mut mock_hdk = MockHdkT::new();
+        
+        let zome_info = fixt!(ZomeInfo);
+        // do these mock functions need to be called in the same order as the actual code?
+        // init calls create_receive_signal_cap_grant, which calls zome_info and create_cap_grant
+        mock_hdk
+            .expect_zome_info()
+            .times(1)
+            .return_const(Ok(zome_info.clone())); // define what it should return when it matches the `with` statement, or matches at all (without `.with`)
+
+        // create_cap_grant calls just `create` under the hood
+        let mut functions: GrantedFunctions = BTreeSet::new();
+        functions.insert((zome_info.zome_name, "recv_remote_signal".into()));
+        // expected is for the .with, and is b/c create parameter is of type EntryWithDefId
+        let expected = EntryWithDefId::new(
+            EntryDefId::CapGrant,
+            Entry::CapGrant(CapGrantEntry {
+                tag: "".into(),
+                access: ().into(),
+                functions,
+            }),
+        );
+        let header_hash = fixt!(HeaderHash);
+        mock_hdk
+            .expect_create()
+            .with(mockall::predicate::eq(expected))
+            .times(1)
+            .return_const(Ok(header_hash.clone()));
+        
+        // init also calls join_project_during_init, drilling down, it calls the following hdk functions under the hood
+        // TODO: handle cases where a path with 1 or more parents doesn't yet exist in the DHT (this would call `create` and `create_link` additional times)
+        let entry_hash = fixt!(EntryHash);
+        mock_hdk
+            .expect_hash_entry()
+            .times(3)
+            .return_const(Ok(entry_hash.clone()));
+
+        let expected_output = fixt!(Element);
+        let exp_vec = vec![Some(expected_output)];
+        mock_hdk
+            .expect_get()
+            .times(1)
+            .return_const(ExternResult::Ok(exp_vec));
+
+        let agent_info = fixt!(AgentInfo);
+
+        mock_hdk
+            .expect_agent_info()
+            .times(1)
+            .return_const(Ok(agent_info.clone()));
+        
+        mock_hdk
+            .expect_create()
+            .times(1)
+            .return_const(Ok(header_hash.clone()));
+        
+        mock_hdk
+            .expect_create_link()
+            .times(1)
+            .return_const(Ok(header_hash));
+
+        set_hdk(mock_hdk);
+
+        let result = init(());
+        assert_eq!(result.is_ok(), true);
+    }
+}

--- a/dna/tests/src/project/project_init.rs
+++ b/dna/tests/src/project/project_init.rs
@@ -11,19 +11,25 @@ pub mod tests {
     use projects::project::project_meta::crud::ProjectMeta;
     use projects::project::project_meta::validate::*;
     use projects::init;
+    use projects::project::member::entry::Member;
+    use assert_matches::assert_matches;
 
+    /// Unit testing the `init()` function that is called to ensure an agent can receive signals
+    /// and is listed as a member in the project. There are two main functions called within `init()`: `create_receive_signal_cap_grant()`
+    /// and `join_project_during_init()`. The first function creates a capabilities grant token
+    /// allowing other agents to call a specified zome function (used for sending signals for responsive UI updates)
+    /// and the second function creates a member entry and links it to MEMBER_PATH 
     #[test]
     fn test_init() {
                
         let mut mock_hdk = MockHdkT::new();
         
         let zome_info = fixt!(ZomeInfo);
-        // do these mock functions need to be called in the same order as the actual code?
         // init calls create_receive_signal_cap_grant, which calls zome_info and create_cap_grant
         mock_hdk
             .expect_zome_info()
             .times(1)
-            .return_const(Ok(zome_info.clone())); // define what it should return when it matches the `with` statement, or matches at all (without `.with`)
+            .return_const(Ok(zome_info.clone())); 
 
         // create_cap_grant calls just `create` under the hood
         let mut functions: GrantedFunctions = BTreeSet::new();
@@ -45,43 +51,71 @@ pub mod tests {
             .return_const(Ok(header_hash.clone()));
         
         // init also calls join_project_during_init, drilling down, it calls the following hdk functions under the hood
-        // TODO: handle cases where a path with 1 or more parents doesn't yet exist in the DHT (this would call `create` and `create_link` additional times)
-        // 1. `Path::from(MEMBER_PATH).ensure()?;`
-        // 2. `let member_path_address = Path::from(MEMBER_PATH).hash()?;`
-        // 3. `let member_entry_hash = hash_entry(&member)?;`
-        let entry_hash = fixt!(EntryHash);
+        // TODO: test when the MEMBER_PATH doesn't exist in DHT
+        let member_path = Path::from("member");
+        let member_path_entry = Entry::try_from(member_path).unwrap();
+        let member_path_entry_hash = fixt!(EntryHash);
         mock_hdk
-            .expect_hash_entry()
-            .times(3)
-            .return_const(Ok(entry_hash.clone()));
-
-        // relates to `Path::from(MEMBER_PATH).ensure()?;`
-        let expected_output = fixt!(Element);
-        let exp_vec = vec![Some(expected_output)];
-        mock_hdk
-            .expect_get()
+            .expect_hash_entry() // called from `Path::from(MEMBER_PATH).ensure()?;`
+            .with(mockall::predicate::eq(member_path_entry.clone()))
             .times(1)
-            .return_const(ExternResult::Ok(exp_vec));
+            .return_const(Ok(member_path_entry_hash.clone()));
 
+
+        let member_path_get_input = vec![GetInput::new(
+            AnyDhtHash::from(member_path_entry_hash.clone()), GetOptions::content()
+        )];
+        // assuming the path exists on DHT
+        let expected_get_output = vec![Some(fixt!(Element))];
+        mock_hdk
+            .expect_get() // called from `Path::from(MEMBER_PATH).ensure()?;`
+            .with(mockall::predicate::eq(member_path_get_input))
+            .times(1)
+            .return_const(Ok(expected_get_output));
+
+        mock_hdk
+            .expect_hash_entry() // called from `let member_path_address = Path::from(MEMBER_PATH).hash()?;`
+            .with(mockall::predicate::eq(member_path_entry.clone()))
+            .times(1)
+            .return_const(Ok(member_path_entry_hash.clone()));
+        
         let agent_info = fixt!(AgentInfo);
         mock_hdk
             .expect_agent_info()
             .times(1)
             .return_const(Ok(agent_info.clone()));
         
+        let member = Member {
+            address: WrappedAgentPubKey(agent_info.agent_initial_pubkey),
+        };
+        let create_member_input = EntryWithDefId::try_from(member.clone()).unwrap();
         mock_hdk
             .expect_create()
+            .with(mockall::predicate::eq(create_member_input))
             .times(1)
             .return_const(Ok(header_hash.clone()));
         
+        let member_hash = fixt!(EntryHash);
+        let member_entry = Entry::try_from(member.clone()).unwrap();
+        mock_hdk
+            .expect_hash_entry() // called from `let member_entry_hash = hash_entry(&member)?;`
+            .with(mockall::predicate::eq(member_entry))
+            .times(1)
+            .return_const(Ok(member_hash.clone()));
+        let create_link_input = CreateLinkInput::new(
+            member_path_entry_hash,
+            member_hash,
+            LinkTag::from(()),
+        );
         mock_hdk
             .expect_create_link()
+            .with(mockall::predicate::eq(create_link_input))
             .times(1)
             .return_const(Ok(header_hash));
 
         set_hdk(mock_hdk);
 
         let result = init(());
-        assert_eq!(result.is_ok(), true);
+        assert_matches!(result, Ok(InitCallbackResult::Pass));
     }
 }

--- a/dna/tests/src/project/project_init.rs
+++ b/dna/tests/src/project/project_init.rs
@@ -46,12 +46,16 @@ pub mod tests {
         
         // init also calls join_project_during_init, drilling down, it calls the following hdk functions under the hood
         // TODO: handle cases where a path with 1 or more parents doesn't yet exist in the DHT (this would call `create` and `create_link` additional times)
+        // 1. `Path::from(MEMBER_PATH).ensure()?;`
+        // 2. `let member_path_address = Path::from(MEMBER_PATH).hash()?;`
+        // 3. `let member_entry_hash = hash_entry(&member)?;`
         let entry_hash = fixt!(EntryHash);
         mock_hdk
             .expect_hash_entry()
             .times(3)
             .return_const(Ok(entry_hash.clone()));
 
+        // relates to `Path::from(MEMBER_PATH).ensure()?;`
         let expected_output = fixt!(Element);
         let exp_vec = vec![Some(expected_output)];
         mock_hdk
@@ -60,7 +64,6 @@ pub mod tests {
             .return_const(ExternResult::Ok(exp_vec));
 
         let agent_info = fixt!(AgentInfo);
-
         mock_hdk
             .expect_agent_info()
             .times(1)

--- a/dna/tests/src/project/project_init.rs
+++ b/dna/tests/src/project/project_init.rs
@@ -21,9 +21,22 @@ pub mod tests {
     /// and the second function creates a member entry and links it to MEMBER_PATH 
     #[test]
     fn test_init() {
-               
         let mut mock_hdk = MockHdkT::new();
+
+        // make a mutable reference to mock_hdk so that it can be mutated with each function
+        let mock_hdk_ref = &mut mock_hdk;
         
+        // setup a mock of `create_receive_signal_cap_grant` because it is called by `init`
+        setup_create_receive_signal_cap_grant_mock(mock_hdk_ref);
+        
+        // setup a mock of `join_project_during_init` because it is called by `init`
+        setup_join_project_during_init_mock(mock_hdk_ref);
+
+        set_hdk(mock_hdk);
+        let result = init(());
+        assert_matches!(result, Ok(InitCallbackResult::Pass));
+    }
+    fn setup_create_receive_signal_cap_grant_mock(mock_hdk: &mut MockHdkT) {
         let zome_info = fixt!(ZomeInfo);
         // init calls create_receive_signal_cap_grant, which calls zome_info and create_cap_grant
         mock_hdk
@@ -49,8 +62,8 @@ pub mod tests {
             .with(mockall::predicate::eq(expected))
             .times(1)
             .return_const(Ok(header_hash.clone()));
-        
-        // init also calls join_project_during_init, drilling down, it calls the following hdk functions under the hood
+    }
+    fn setup_join_project_during_init_mock(mock_hdk: &mut MockHdkT) {
         // TODO: test when the MEMBER_PATH doesn't exist in DHT
         let member_path = Path::from("member");
         let member_path_entry = Entry::try_from(member_path).unwrap();
@@ -88,6 +101,7 @@ pub mod tests {
         let member = Member {
             address: WrappedAgentPubKey(agent_info.agent_initial_pubkey),
         };
+        let header_hash = fixt!(HeaderHash);
         let create_member_input = EntryWithDefId::try_from(member.clone()).unwrap();
         mock_hdk
             .expect_create()
@@ -112,10 +126,5 @@ pub mod tests {
             .with(mockall::predicate::eq(create_link_input))
             .times(1)
             .return_const(Ok(header_hash));
-
-        set_hdk(mock_hdk);
-
-        let result = init(());
-        assert_matches!(result, Ok(InitCallbackResult::Pass));
-    }
+    }    
 }

--- a/dna/tests/src/project/project_init.rs
+++ b/dna/tests/src/project/project_init.rs
@@ -2,33 +2,33 @@
 pub mod tests {
     use crate::project::fixtures::fixtures::ProjectMetaFixturator;
     use ::fixt::prelude::*;
+    use assert_matches::assert_matches;
     use hdk::prelude::*;
     use hdk_crud::WrappedAgentPubKey;
     use holochain_types::prelude::option_entry_hashed;
     use holochain_types::prelude::ElementFixturator;
     use holochain_types::prelude::ValidateDataFixturator;
+    use projects::init;
     use projects::project::error::Error;
+    use projects::project::member::entry::Member;
     use projects::project::project_meta::crud::ProjectMeta;
     use projects::project::project_meta::validate::*;
-    use projects::init;
-    use projects::project::member::entry::Member;
-    use assert_matches::assert_matches;
 
     /// Unit testing the `init()` function that is called to ensure an agent can receive signals
     /// and is listed as a member in the project. There are two main functions called within `init()`: `create_receive_signal_cap_grant()`
     /// and `join_project_during_init()`. The first function creates a capabilities grant token
     /// allowing other agents to call a specified zome function (used for sending signals for responsive UI updates)
-    /// and the second function creates a member entry and links it to MEMBER_PATH 
+    /// and the second function creates a member entry and links it to MEMBER_PATH
     #[test]
     fn test_init() {
         let mut mock_hdk = MockHdkT::new();
 
         // make a mutable reference to mock_hdk so that it can be mutated with each function
         let mock_hdk_ref = &mut mock_hdk;
-        
+
         // setup a mock of `create_receive_signal_cap_grant` because it is called by `init`
         setup_create_receive_signal_cap_grant_mock(mock_hdk_ref);
-        
+
         // setup a mock of `join_project_during_init` because it is called by `init`
         setup_join_project_during_init_mock(mock_hdk_ref);
 
@@ -42,7 +42,7 @@ pub mod tests {
         mock_hdk
             .expect_zome_info()
             .times(1)
-            .return_const(Ok(zome_info.clone())); 
+            .return_const(Ok(zome_info.clone()));
 
         // create_cap_grant calls just `create` under the hood
         let mut functions: GrantedFunctions = BTreeSet::new();
@@ -74,9 +74,9 @@ pub mod tests {
             .times(1)
             .return_const(Ok(member_path_entry_hash.clone()));
 
-
         let member_path_get_input = vec![GetInput::new(
-            AnyDhtHash::from(member_path_entry_hash.clone()), GetOptions::content()
+            AnyDhtHash::from(member_path_entry_hash.clone()),
+            GetOptions::content(),
         )];
         // assuming the path exists on DHT
         let expected_get_output = vec![Some(fixt!(Element))];
@@ -91,13 +91,13 @@ pub mod tests {
             .with(mockall::predicate::eq(member_path_entry.clone()))
             .times(1)
             .return_const(Ok(member_path_entry_hash.clone()));
-        
+
         let agent_info = fixt!(AgentInfo);
         mock_hdk
             .expect_agent_info()
             .times(1)
             .return_const(Ok(agent_info.clone()));
-        
+
         let member = Member {
             address: WrappedAgentPubKey(agent_info.agent_initial_pubkey),
         };
@@ -108,7 +108,7 @@ pub mod tests {
             .with(mockall::predicate::eq(create_member_input))
             .times(1)
             .return_const(Ok(header_hash.clone()));
-        
+
         let member_hash = fixt!(EntryHash);
         let member_entry = Entry::try_from(member.clone()).unwrap();
         mock_hdk
@@ -116,15 +116,12 @@ pub mod tests {
             .with(mockall::predicate::eq(member_entry))
             .times(1)
             .return_const(Ok(member_hash.clone()));
-        let create_link_input = CreateLinkInput::new(
-            member_path_entry_hash,
-            member_hash,
-            LinkTag::from(()),
-        );
+        let create_link_input =
+            CreateLinkInput::new(member_path_entry_hash, member_hash, LinkTag::from(()));
         mock_hdk
             .expect_create_link()
             .with(mockall::predicate::eq(create_link_input))
             .times(1)
             .return_const(Ok(header_hash));
-    }    
+    }
 }

--- a/dna/zomes/projects/src/lib.rs
+++ b/dna/zomes/projects/src/lib.rs
@@ -15,7 +15,7 @@ use project::{
 };
 
 #[hdk_extern]
-fn init(_: ()) -> ExternResult<InitCallbackResult> {
+pub fn init(_: ()) -> ExternResult<InitCallbackResult> {
   // authorize receive_signal
   // not the issue, takes about 2 ms
   create_receive_signal_cap_grant()?;

